src/translator/quality_estimator.cpp

```diff
@@ -0,0 +1,133 @@
+            if (words[wordIdx].toString()==words[wordIdx].DEFAULT_EOS_ID.toString()){
+                break;
+            }
+            ByteRange subword = sentence.target.wordAsByteRange(sentenceIdx, wordIdx);
+            size_t subword_begin = subword.begin;
+            size_t subword_end = subword.end;
+            char first_word = sentence.target.text[subword_begin];
+            wordIdx++;
+            if (first){
+                first = false;
+                previous_p = p;
+                map[std::make_tuple(subword_begin, subword_end)] = previous_p;
+                continue;
+            }
+            if (first_word == ' '){
+                target_words.push_back(map);
+                map.clear();
+                map[std::make_tuple(subword_begin+1, subword_end)] = p;
+            }
+            else{
+                map[std::make_tuple(subword_begin, subword_end)] = p;
+            }
+        }
+        //last word
+        if (!map.empty()){
+            target_words.push_back(map);
+        }
+    }
+    this->mapping = target_words;
+}
+ intgemm::AlignedVector<QualityEstimator::FeatureVector> QualityEstimator::extractFeatures(Response& sentence){
+     auto quality_scores = sentence.qualityScores[0]; //this is expected to be only one
+     auto map = this->mapping;
+     const intgemm::Index n_rows = map.size();
+     const intgemm::Index width = 64;
+     intgemm::AlignedVector<QualityEstimator::FeatureVector> feature_matrix(n_rows * width);
+     intgemm::Index current_idx = 0;
+     for(auto &word: map) {
+        bool first=true;
+        size_t first_subword_idx;
+        size_t last_subword_idx;
+        float bpe_scores_sum = 0;
+        float bpe_scores_min;
+        for (auto &subword:word){
+            auto tmp = subword.first;
+            float bpe_score = subword.second;
+            if (first){
+                bpe_scores_min = bpe_score;
+                first_subword_idx = std::get<0>(tmp);
+                first=false;
+            }
+            if(bpe_score<bpe_scores_min){
+                bpe_scores_min = bpe_score;
+            }
+            last_subword_idx = std::get<1>(tmp);
+            bpe_scores_sum += bpe_score;
+        }
+        float bpe_mean = bpe_scores_sum/word.size();
+        float model_scores = quality_scores.sequence/quality_scores.word.size();
+        auto features = QualityEstimator::FeatureVector(bpe_mean,bpe_scores_min,word.size(),model_scores);
+        feature_matrix[current_idx] = features;
+        current_idx +=1;
+        int subword_len = last_subword_idx - first_subword_idx;
+        auto final_word = sentence.target.text.substr(first_subword_idx, subword_len);
+     }
+    // we still need to subtract mean and divide by std
+     return feature_matrix;
+ }
+
+
+void QualityEstimator::compute_quality_scores(Response& sentence, Words words) {
+    this->mapBPEToWords(sentence, words);
+    auto feature_vector = this->extractFeatures(sentence);
```

Comment:
I'm guessing this is the work in progress part since we haven't used the features yet.